velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,497 @@
+    SCOPED_TRACE(fmt::format(
+        "Input: {}\nOptimized: {}\nExpected: {}",
+        expression->toString(),
+        optimized->toString(),
+        expected->toString()));
+    ASSERT_TRUE(*optimized == *expected);
+
+    const auto data = fuzzFlat(type);
+    const auto optimizedResult = evaluate(optimized, data);
+    const auto expectedResult = evaluate(expected, data);
+    test::assertEqualVectors(optimizedResult, expectedResult);
+  }
+
+  /// Validates expression optimization and provides an option to specify input
+  /// types for lambda expressions in the input and expected expressions.
+  /// @param input Input expression to be optimized.
+  /// @param expected Expected expression after optimization.
+  /// @param type Type of input and expected expressions.
+  /// @param lambdaInputTypes Types of inputs to lambda expression that could be
+  /// present in 'input'.
+  /// @param expectedLambdaInputTypes Types of inputs to lambda expression that
+  /// could be present in 'expected'.
+  void testExpression(
+      const std::string& input,
+      const std::string& expected,
+      const RowTypePtr& type = ROW({}),
+      const std::vector<TypePtr>& lambdaInputTypes = {},
+      const std::vector<TypePtr>& expectedLambdaInputTypes = {}) {
+    const auto expr = makeTypedExpr(input, type, lambdaInputTypes);
+    const auto expectedExpr =
+        makeTypedExpr(expected, type, expectedLambdaInputTypes);
+    testExpression(expr, expectedExpr, type);
+  }
+
+  /// Validates that the subexpressions that throw an exception during
+  /// optimization are replaced with a fail expression when optimize is called
+  /// with 'replaceEvalErrorWithFailExpr = true'. Also validates that the
+  /// optimized expression containing fail expression(s) will throw the expected
+  /// error upon evaluation.
+  /// @param input Input expression to be optimized.
+  /// @param expected Expected expression as a string.
+  /// @param error Expected error message upon evaluating input expression.
+  /// @param type Type of input expression.
+  /// @param lambdaInputTypes Types of inputs to lambda expression that could be
+  /// present in 'input'.
+  void assertFailCall(
+      const std::string& input,
+      const std::string& expected,
+      const std::string& error,
+      const RowTypePtr& type = ROW({}),
+      const std::vector<TypePtr>& lambdaInputTypes = {}) {
+    const auto expr = makeTypedExpr(input, type, lambdaInputTypes);
+    auto optimized = expression::optimize(expr, queryCtx_.get(), pool(), true);
+    ASSERT_EQ(optimized->toString(), expected);
+    VELOX_ASSERT_THROW(evaluate(optimized, fuzzFlat(type)), error);
+  }
+
+  void setQueryTimeZone(const std::string& timeZone) {
+    queryCtx_->testingOverrideConfigUnsafe(
+        {{core::QueryConfig::kSessionTimezone, timeZone},
+         {core::QueryConfig::kAdjustTimestampToTimezone, "true"}});
+  }
+
+ private:
+  core::TypedExprPtr makeTypedExpr(
```

Comment:
Add a comment for this method. Please add that you will use the DuckDB parser for parsing the input text parameter.